File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] freq = new int[26];
        int start = 0, max = 0, win = Integer.MIN_VALUE;
        for(int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            freq[ch - 'A']++;
            max = Math.max(max,freq[ch - 'A']);
            if((i - start + 1) - max > k) {
                    freq[s.charAt(start) - 'A']--;
                    start++;
            }
            win = Math.max(win, (i - start + 1));
        }
        return win;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution(object):
    def characterReplacement(self, s, k):
        map = {}
        left, max_length, max_frequency = 0, 0, 0
        for right in range(len(s)):
            c = s[right]
            map[c] = map.get(c,0) + 1
            max_frequency = max(max_frequency,map[c])

            length = right - left + 1
            if length - max_frequency > k:
                ch = s[left]
                if map[ch] == 1:
                    map.pop(ch)
                else:
                    map[ch] = map[ch] - 1
                left += 1
            max_length = max(max_length,right - left + 1)

        return max_length
        